docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+
+**Pods with limits specified are not very well protected from pods without limits specified**. If a pod without an explicit memory limit goes overboard and uses lots of memory, other pods (with specified memory limits) might be killed. This is bad, because users often want a way to launch containers that have resources guarantees, and that stay up reliably.
+
+On the other hand, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Pods often don’t use all the resources that they request which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a node with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in more pods onto the node, but Kubernetes will not schedule new pods on the node.
+
+This proposal provides mechanisms for solving both the reliability and efficiency problems, by allowing pods to specify levels of resource guarantees. Containers will be able to *request* for a minimum resource guarantee. The *request* is different from the *limit* - containers will not be allowed to exceed resource limits.
+
+Additionally, we will support 3 different tiers of pods. Guaranteed pods (where requests and limits are equal) will receive the highest priority and will get all the resources they need. Burstable pods, which specify request < limit will be guaranteed their request and can “burst” and use more resources when available. Best-Effort pods, which don’t specify a request, can use resources only if not being used by other pods. In Borg, supporting best-effort pods increased utilization by about 20%, and we hope to see similar improvements in Kubernetes.
+
+## QoS Functionality
+
+For each resource, containers can specify a resource request and limit, 0 <= request <= limit <= Infinity. If the container is successfully scheduled, the container is guaranteed the amount of resource requested. The container will not be allowed to exceed the specified limit. How the request and limit are enforced depends on whether the resource is [compressible or incompressible](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/resources.md).
+
+### Compressible Resource Guarantees
+- For now, we are only supporting CPU.
+- Jobs that don’t have a request specified, will have a request of 0.
+- Containers are guaranteed to get the amount of CPU they request, they may or may not get additional CPU time (depending on the other jobs running).
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 60% of the CPU, and container B requests for 30% of the CPU. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10% of CPU will be distributed to A and B in a 2:1 ratio.

```

Comment:
Implementation for this was discussed below - let me know if you see any issues with the implementation!